RFS/src/main/java/org/opensearch/migrations/bulkload/lucene/LuceneDocumentsReader9.java

```diff
@@ -0,0 +1,230 @@
+import lombok.extern.slf4j.Slf4j;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+import shadow.lucene9.org.apache.lucene.document.Document;
+import shadow.lucene9.org.apache.lucene.index.DirectoryReader;
+import shadow.lucene9.org.apache.lucene.index.LeafReaderContext;
+import shadow.lucene9.org.apache.lucene.index.SoftDeletesDirectoryReaderWrapper;
+import shadow.lucene9.org.apache.lucene.index.StoredFields;
+import shadow.lucene9.org.apache.lucene.store.FSDirectory;
+import shadow.lucene9.org.apache.lucene.util.BytesRef;
+
+@RequiredArgsConstructor
+@Slf4j
+public class LuceneDocumentsReader9 implements LuceneDocumentsReader {
+
+    protected final Path indexDirectoryPath;
+    protected final boolean softDeletesPossible;
+    protected final String softDeletesField;
+ 
+    /**
+     * There are a variety of states the documents in our Lucene Index can be in; this method extracts those documents
+     * that would be considered "live" from the ElasticSearch/OpenSearch perspective.  The most important thing to know is
+     * that Lucene segments are immutable.  For additional context, it is highly recommended to read this section of the
+     * Lucene docs for a high level overview of the topics involved:
+     *
+     * https://lucene.apache.org/core/8_0_0/core/org/apache/lucene/codecs/lucene80/package-summary.html
+     *
+     * When ElasticSearch/OpenSearch deletes a document, it doesn't actually remove it from the Lucene Index.  Instead, what
+     * happens is that the document is marked as "deleted" in the Lucene Index, but it is still present in the Lucene segment
+     * on disk.  The next time a merge occurs, that segment will be deleted, and the deleted documents in it are thereby
+     * removed from the Lucene Index.  A similar thing happens when a document is updated; the old document is marked as
+     * "deleted" in the Lucene segment and the new version of the document is added in a new Lucene segment.  Until a merge
+     * occurs, both the old and new versions of the document will exist in the Lucene Index in different segments, though only
+     * the new version will be returned in search results.  This means that from an ES/OS perspective, you could have a single
+     * document that has been created, deleted, recreated, updated, etc. multiple times at the Elasticsearch/OpenSearch level
+     * and only a single version of the doc would exist when you queried the ES/OS Index - but every single iteration of that
+     * document might still exist in the Lucene segments on disk, all of which have the same _id (from the ES/OS perspective).
+     *
+     * Additionally, Elasticsearch 7 introduced a feature called "soft deletes" which allows you to mark a document as
+     * "deleted" in the Lucene Index without actually removing it from the Lucene Index.  From what I can gather, soft deletes
```

Comment:
Instead of "From what I gather" - you can reference the link below, as you do, and just say that it's used for certain shard synchronization or management issues, then cite the link.  Farther below, it's worth calling out that soft-deleted docs WILL be eventually FULLY-deleted lucene docs & that's why we don't want to load them onto the target.